Added Mixin tutorial + 1 more (#109)
FabricMC/fabric-docs: develop/using-mixins.md

---

> title: Using Mixins

Reviewer: Having the title just `Mixins` may be more preferable, as titles of other pages in this section don't use verbs

When creating a mixin, you must understand exactly where the code is located which you want to alter or add to. For this example, we will be making arrows explode when they hit entities.
Following yarn naming conventions, we can infer that the code we are seeking to inject into will be in the `ArrowEntity` class.
We still need to check that this is actually the right class for what we want to accomplish, so it is necessary to [view the class's source code](https://fabricmc.net/wiki/tutorial:reading_mc_code).

Upon inspection, there is an `onHit(LivingEntity target)` method, so this is a good candidate.

Now that we know where we want to inject, we can start writing a mixin. Start by locating the mixin package, `java/com/example/docs/mixin` in the reference mod. Normally, mixins are named after their target class, so in this case ours would be `ArrowEntityMixin`. Add the `@Mixin(SomeClass.class)` annotation, and add it to the mixin manifest, as follows.

:::code-group
```java [ArrowEntityMixin.java]
@Mixin(ArrowEntity.class)
public class ArrowEntityMixin {
}
```
```json [modid.mixins.json]
{
  ...,
  "mixins": [
    "ArrowEntityMixin"
  ],
  ...
}
```
:::

> At the moment, we've targetted the class but haven't actually made it do anything. For this we need a method annotated with `@Inject`, as follows.

Reviewer: ```suggestion
At the moment, we've targeted the class but haven't actually made it do anything. For this we need a method annotated with `@Inject`, as follows.
```

> At the moment, we've targetted the class but haven't actually made it do anything. For this we need a method annotated with `@Inject`, as follows.

Reviewer: > For this we need a method annotated with `@Inject`, as follows.

This sentence needs to emphasize that `@Inject` is used here *for the purpose of this example,* and that other annotations are used where applicable

@[code lang=java transcludeWith=:::1](@/reference/latest/src/main/java/com/example/docs/mixin/ArrowEntityMixin.java)

This may seem complicated at first, but it is relatively simple once you break it down into its components.

> - `at = @At("TAIL")`: Every injection requires an injection point, possible values include, but are not limited to `HEAD`, `TAIL`, `INVOKE` and `RETURN`. `TAIL` selects the end of the method, regardless of where it returns. In this case, returns do not matter, since the target method never calls `return;`.

Reviewer: ```suggestion
- `at = @At("TAIL")`: Every injection requires an injection point, possible values include, but are not limited to `HEAD`, `TAIL`, `INVOKE` and `RETURN`. `TAIL` selects the end of the method, regardless of where it returns. In this case, returns do not matter, since the target method never uses `return;`.
```

- `method = "onHit"`: Since the method is called `onHit`, and we want to inject into it, we write `onHit` in the method parameter. As it is the only method called `onHit`, it isn't necessary to specify its parameters, but when it is, the plugins' suggestions will help you a lot.

> - `LivingEntity target, CallbackInfo ci`: Every inject-annotated method is passed a `CallbackInfo` object, which can be used to return methods when the injected method has the `cancellable = true` parameter. It also passes the target method's parameter, a `LivingEntity`, which is used to summon the explosion.

Reviewer: ```suggestion
- `LivingEntity target, CallbackInfo ci`: Every inject-annotated method is passed a `CallbackInfo` object, which can be used to return from methods when the injection has the `cancellable = true` parameter. It also passes the target method's parameter, a `LivingEntity`, which is used to summon the explosion.
```

> - `target.getY() + (target.getHeight() / 2)`: Entity y position is below the feet, by adding half the entity's height, the explosion is summoned in the middle of the entity.

Reviewer: ```suggestion
- `target.getY() + (target.getHeight() / 2)`: Entities' Y position is located at their feet level. We make the explosion appear in the middle of the entity by adding half the entity's height.
```

> As we can see by testing the mod, arrows now detonate an explosion upon hitting an entity.

Reviewer: ```suggestion
As we can see by testing the mod, arrows now detonate upon hitting an entity.
```
or
```suggestion
As we can see by testing the mod, arrows now summon an explosion upon hitting an entity.
```
